### DISCUSSION TITLE
Improve DX of adding modules with CLI

### DISCUSSION THREAD

**Discussion by pi0** [nuxt/nuxt#17785] [RFCs]
Adding common modules is a tedious task especially when you want to make a starter project. There are several ways they can be added:

- The `create-nuxt-app` approach asks to install them when creating from a template
- Auto-discovery if they are installed in package.json
- Adding with CLI `nuxi module add @nuxtjs/color-mode`

Using template, is the most common way right now but it has disadvantages:
- Creating a new project takes more time
- Users might not know yet what they need
- Requires templating logic if module addition is not enough

Using auto-discovery is probably the sexiest possible way but it has disadvantages:
- There might be installed dependencies that we do not want as a module
- Needs checking each dependency is a module or not
- We lose the single source of trust about nuxt config

Using CLI:
- Solves template limitation but allowing to do it on demand
- Uses explicit dependency and config (nuxtrc or nuxt.config) 
- Allows module authors to define custom (postinstall, postupgrade) hooks by themselves
 
One more idea I thought with #570, is that we can offer auto install by detecting usage (adding `pwa` key to `nuxt.config` or an auto import detected without module). Probably another RFC but worth exploring...

----

Notes:

- Legacy RFC: https://github.com/nuxt/rfcs/issues/31

**Comment by atinux** [comment]
I agree that the CLI is the best way for the long run, maintaining `create-nuxt-app` is already a tedious task and it does not include all the possibilities.

Offering auto-install will be the ice on the cake once we have a strong CLI with `nuxi` to communicate with the user nuxt config and project dependencies.

**Reply by Intevel** [reply]
This method has advantages, but I think a comprehensive create-nuxt-app command is good, maybe with example setups for the selectable options.
